design/secrets_provider_milestone1_2_design.md

```diff
@@ -0,0 +1,380 @@
+| 4    | *Outside trigger,* retrieve secrets when notified by websocket connection | Websocket connection is opened and when there is a change on the server, it will send update | - Efficient because only do the job when really needed<br />- Immediate update of K8s Secret on secret change in Conjur | - Need development in Conjur side to support notification mechanism to notify only necessary secrets providers<br />- Very expensive to hold a socket open just for few rare notifications<br />- Significant development required because it requires a new event mechanism on server |
+
+Decision: **Solution #2, Outside trigger using pub/sub message
+**
+
+The Pub/Sub pattern is a pattern where senders publish messages without knowing who is subscribing to those messages. Once an event is emitted or a message is published, all receivers that subscribe to that event will get a notification. For our purposes, we propose to use a 3rd party pub/sub message handlers, ZMQ, the publisher, that will trigger an event whenever a Conjur secret has been updated. The Secrets Provider, the subscriber, will listen on a port for these events/messages. Once a message is received this will prompt the Secrets Provider to update the K8s Secrets with the updated Conjur Secrets.
+
+We chose to use the pub/sub pattern over waiting for a configurable amount of time because this delivers a nicer customer experience. The customer's K8s Secrets will be updated immediately which in turn avoids situations where customer applications fail to connect to their endpoints. 
+
+#### Flow
+
+1. Configure K8s Secrets with ***conjur-map\*** metadata defined 
+
+   - *Optional:* add label to K8s Secret 
+
+2. Create Service Account for Secrets Provider 
+
+3. Create Role / ClusterRole for the Service Accounts with **get**/**update** privileges
+
+4. Create RoleBinding / ClusterRoleBinding and bind the Service Accounts to the role from previous step
+
+5. Create/Deploy Secrets Providers as Pods in deployment manifest(s) **(\*Milestone 2)\***
+
+6. - If label filtering is defined, add "K8S_SECRET_LABEL" in the Secrets Provider manifest(s)
+
+7. When the Secrets Provider receives a message that Conjur secrets have changed, the process of fetching and updating the K8s Secret(s) repeats
+
+
+
+#### Customer experience
+
+Because the Secrets Provider needs to now update K8s Secrets with Conjur Secrets continuously, it can no longer run as a Job and will require that the customer deploy the Secrets Provider as a Pod (kind: Pod instead of kind: Job)
```

Comment:
it can also be a Deployment, DeploymentConfig or StatefulSet as today

Reply:
As discussed, when using Deployment to spin up a pod and the task finishes, the will container exit and be deleted. Now that the Secrets Provider needs to sit separately, when that container exists so too will the pod. Therefore, when using Deployment, Kuberentes will try to restart the Pod. After a few attempted restarts, the pod will error in a BackOff state. Because of this, we are unable to use Deployment to spin up a pod and must use a Job.